{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Menu(infile):\n",
    "    \n",
    "    \"\"\" This funtion displays the menu and then calls the Mainmenu function that \n",
    "        allows the user to select the displayed options.\n",
    "        Arguments: pdb infile (that gets loaded once it is read)\n",
    "    \"\"\"\n",
    "    \n",
    "#     from Mainmenu import Mainmenu\n",
    "    \n",
    "    space = (\" \")\n",
    "    a = \" PDB ANALYSIS INTERFACE \"\n",
    "    b = \" Pick your choice from below \"\n",
    "    c = \" [R] = Enter name of a PDB file \"\n",
    "    d = \" [S] = Search for PDB files \"\n",
    "    e = \" [W] = Write out PDB files (Files in existence) \"\n",
    "    f = \" [I] = Information on PDB files \"\n",
    "    g = \" [A] = Alignment \"\n",
    "    h = \" [H] = Help \"\n",
    "    i = \" [Q] = Quit \"\n",
    "\n",
    "    # Creating the body head of the program with lines of length 70\n",
    "    ln1 = (space+(\"#\"*70))\n",
    "    ln2 = (space+\"#\"+a+(space*(70-len(a)-2))+\"#\")\n",
    "    ln3 = (space+(\"#\"*70))\n",
    "    ln4 = (space+\"#\"+b+(space*(70-len(b)-2))+\"#\")\n",
    "    ln5 = (space+\"#\"+(space*68)+\"#\")\n",
    "    ln6 = (space+\"#\"+space+c+(space*(70-len(c)-5-35))+(space*30)+space*5+\"#\")\n",
    "    ln7 = (space+\"#\"+space+d+(space*(70-len(d)-5-35))+(space*30)+space*7+\"#\")\n",
    "    ln8 = (space+\"#\"+space+e+(space*(70-len(e)-5-35))+(space*19)+\"#\")\n",
    "    ln9 = (space+\"#\"+space+f+(space*(70-len(f)-5-35))+(space*30)+space*5+\"#\")\n",
    "    ln10 = (space+\"#\"+space+g+(space*(70-len(g)-5-35))+(space*30)+space*7+\"#\")\n",
    "    ln11 = (space+\"#\"+space+h+(space*(70-len(h)-5-35))+(space*30)+space*7+\"#\")\n",
    "    ln12 = (space+\"#\"+space+i+(space*(70-len(i)-5-35))+(space*30)+space*7+\"#\")\n",
    "    ln13 = (space+(\"#\"*70))\n",
    "\n",
    "    \n",
    "    lines = [ln1,ln2,ln3,ln4,ln5,ln6,ln7,ln8,ln9,ln10,ln11,ln12,ln13]\n",
    "    \n",
    "    print(\"\\n\")\n",
    "    # Read all lines to recreate menu\n",
    "    for line in lines:\n",
    "        print(line)\n",
    "    \n",
    "        \n",
    "# #     Selecting an option\n",
    "#     option = input(\"     :\")\n",
    "#     option = option.upper()\n",
    "    \n",
    "    # Initializing the Mainmenu function\n",
    "    Mainmenu(infile)\n",
    "         \n",
    "#     globals()['option'] = input()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Options to analyze PDB files\n",
    "#import sys allows you to use the sys.exit command to quit/logout of the program\n",
    "\n",
    "def Mainmenu(infile):\n",
    "    \n",
    "    \"\"\" This function provides allows the user to select from the available options displayed. \n",
    "        Arguments: option, pdb file\n",
    "    \"\"\"\n",
    "    \n",
    "#     from Read_file import Read_file\n",
    "#     from Search_file import Search_file\n",
    "#     from Write_file import Write_file\n",
    "#     from Infor_file import Infor_file\n",
    "#     from Align_2files import Align_2files\n",
    "#     from Help_file import Help_file\n",
    "    import sys\n",
    "    global path\n",
    "    \n",
    "    m = ''\n",
    "    while m == '':\n",
    "#         print(\"************MAIN MENU**************\")\n",
    "\n",
    "#         print(\"\"\" Available options \n",
    "#                 [R]: Enter name of a PDB file \n",
    "#                 [S]: Search for PDB files\n",
    "#                 [W]: Write out PDB files (PDB files in existence) \n",
    "#                 [I]: Information on PDB files\n",
    "#                 [A]: Alignment\n",
    "#                 [H]: Help\n",
    "#                 [Q]: Quit\n",
    "#             \"\"\")\n",
    "    \n",
    "        choice = input('Please enter your choice: ')\n",
    "        if choice.upper() == 'R':\n",
    "            m = Read_file(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'S':\n",
    "            m = Search_file(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'W':\n",
    "            m = Write_file(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'I':\n",
    "            m = Infor_file(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'A':\n",
    "            m = Align_2files(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'H':\n",
    "            m = Help_file(infile)\n",
    "            Menu(infile)\n",
    "            \n",
    "        elif choice.upper() == 'Q':\n",
    "            print('Program terminated! Goodbye.')\n",
    "            sys.exit(0)\n",
    "            raise SystemExit\n",
    "            \n",
    "        else:\n",
    "            print('You must only enter R, S, W, I, A, H or Q. Please try again')\n",
    "            Menu(infile)\n",
    "    return m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"The module contain the function to enable a user export the pdb file to a different folder with a new name.\"\"\"\n",
    "\n",
    "def Store_file(filepath):\n",
    "    \"\"\"This function exports and stores the current loaded pdb file into a new file in the defined path.\n",
    "    Arguments: User inputs path and the new pdb filename\"\"\"\n",
    "    \n",
    "    import os\n",
    "    out_path = input (\"\"\"\n",
    "     Enter an authentic path to your file destination and the new filename.pdb \n",
    "     :\"\"\")\n",
    "\n",
    "    # Split the path to obtain the file directory and new filename\n",
    "    out_path1 = out_path.split(\"/\")\n",
    "    directory = \"/\".join(out_path1[:-1])\n",
    "    filename = out_path1[-1]\n",
    "\n",
    "    name_tag = False\n",
    "    # Checking if path exists\n",
    "    if os.path.isdir(directory):\n",
    " \n",
    "        extension = filename.split(\".\")\n",
    "\n",
    "        if len(extension) == 1: # checking if the extention is included\n",
    "            filename = filename + \".pdb\"\n",
    "            out_path = directory + \"/\" + filename\n",
    "            name_tag = True\n",
    "\n",
    "        else:\n",
    "            if len(extension) == 2 and extension[-1] == 'pdb': # confirm right extension name\n",
    "                name_tag = True\n",
    "            else:\n",
    "                print(\"Check your output filename extension and name it appropriately\")\n",
    "                Store_file(filepath)\n",
    "                \n",
    "        if os.path.exists(out_path):\n",
    "            print('The file %s exists!' %out_path)\n",
    "            Store_file(filepath)\n",
    "\n",
    "        if name_tag:\n",
    "            #open file for writing and export\n",
    "            with open(out_path, \"w\") as myfile:\n",
    "\n",
    "                pdbfile = open(filepath,\"r\")\n",
    "                for line in pdbfile:\n",
    "                    myfile.write(line) # write to new file\n",
    "\n",
    "                pdbfile.close() \n",
    "\n",
    "                print(\"Successfully exported the pdb file to %s\" %out_path)\n",
    "        \n",
    "    else:\n",
    "        print(\"The directory does not exist \")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Read_file(infile):\n",
    "    \n",
    "    \"\"\" This function loads a pdb file and reads contents for analysis \"\"\"\n",
    "    \n",
    "    import os\n",
    "    global path\n",
    "    \n",
    "    filepath = input('Enter an authentic path to a pdb file: ')\n",
    "    \n",
    "    # Split the path to obtain the file directory and file name\n",
    "    filepath1 = filepath.split('/')\n",
    "    directory = '/'.join(filepath1[:-1])\n",
    "    filename = filepath1[-1]\n",
    "    \n",
    "    # Checking if filepath and file are true\n",
    "    if os.path.isdir(directory):\n",
    "        if os.path.isfile(filepath):\n",
    "            myfile = open(filepath,'r')\n",
    "            line_1 = myfile.readline()\n",
    "            lastline = myfile.readlines()[-1]\n",
    "            myfile.close()\n",
    "            #all pdb files start with header and have a length of 81 and lastline starts with end\n",
    "            if line_1.startswith('HEADER') and len(line_1) == 81 and lastline.startswith('END'): \n",
    "                print('The File %s has been successfully loaded' %filename)\n",
    "                Store_file(filepath)\n",
    "#                 pdb = filename # reassign the name of the pdb file to update on menu\n",
    "            else:\n",
    "                print('This is not a pdb file')\n",
    "                \n",
    "        else:\n",
    "            extension = filename.split('.')\n",
    "            if extension[-1] == 'pdb':\n",
    "                print('The File %s does not exist' %filename)\n",
    "                \n",
    "            else:\n",
    "                print('%s is not a pdb file' %filename)\n",
    "                \n",
    "    else:\n",
    "        print('The directory does not exist!')\n",
    "     \n",
    "    return infile, filepath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/njesh/python-mini-project-JaneNjeri/Notebooks'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "unexpected EOF while parsing (<ipython-input-29-28240296f82b>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-29-28240296f82b>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    Menu('infile'\u001b[0m\n\u001b[0m                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m unexpected EOF while parsing\n"
     ]
    }
   ],
   "source": [
    "Menu('infile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pdbfile_list():\n",
    "    \n",
    "    \"\"\" This function finds all loaded files in the appropriate folder. \"\"\"\n",
    "    \n",
    "    filepath = input('Please enter the path to where exported pdb files are after being loaded')\n",
    "    \n",
    "    import glob, os\n",
    "    os.chdir(filepath)\n",
    "    file_list = []\n",
    "    for file in glob.glob('*.pdb'):\n",
    "        file_list.append(file)\n",
    "    return file_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Get_seqres():\n",
    "    \n",
    "    \"\"\" This function extracts seqres sequences in a pdb file and outputs it to a text file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    \n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    seq_list = []\n",
    "    with open(filepath,'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('SEQRES'):\n",
    "                line_split = line.split()[4]\n",
    "                seq_list.append(line_split)\n",
    "    \n",
    "    choice1 = input('Enter the name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.write(i)\n",
    "        print('Sequences successfully written!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def glyc_sites():\n",
    "    \n",
    "    \"\"\" This function finds glycosylation sites given G or Y or L then A then P or F or W the TLVGMI as motifs\n",
    "    given an amino acid sequence file. \"\"\"\n",
    "    \n",
    "    import re\n",
    "    import os\n",
    "    filepath = input('Enter the path of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath, 'r') as file:\n",
    "        for i in file:\n",
    "            pattern = re.compile(r'[GYL]A[PFW][TLVGMI]')\n",
    "            matches = pattern.finditer(i)\n",
    "            for match in matches:\n",
    "                print(match)     \n",
    "    return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_glycSites():\n",
    "    \n",
    "    \"\"\" This function checks the glycosylation sites of a protein sequence in a given file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath, 'r') as myfile:\n",
    "        for i in myfile:\n",
    "            glyc_sites(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Search_file(infile):\n",
    "    \n",
    "    \"\"\" This function enables the user to search PDB files that are already loaded, for analysis.\n",
    "        It also, reports glycosylation sites in the protein with a given pattern for the motif.\n",
    "    \"\"\"\n",
    "        \n",
    "    import os\n",
    "    import sys\n",
    "    import re\n",
    "#     pdbfile_list()\n",
    "\n",
    "    #loads all loaded pdb files\n",
    "    file_list = pdbfile_list()\n",
    "    \n",
    "    items = pdbfile_list()\n",
    "    \n",
    "    file_list = [name for name in items if name.endswith('.pdb')]\n",
    "\n",
    "    for count, fileName in enumerate(file_list, 0):\n",
    "        sys.stdout.write(\"[%d] %s\\n\\r\" % (count, fileName))\n",
    "\n",
    "    choice = int(input(\"Select pdb file[0-%s]: \" % count))\n",
    "    print(file_list[choice])\n",
    "            \n",
    "#     choice = infile\n",
    "\n",
    "    #get seqres sequences from the file and outputs to an outfile\n",
    "    with open(choice, 'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line.startswith('SEQRES'):\n",
    "                line_split = line.split()[4]\n",
    "                seq_list.append(line_split)\n",
    "                \n",
    "        for i in seq_list:\n",
    "            pattern = re.compile(r'[GYL]A[PFW][TLVGMI]')\n",
    "            matches = pattern.finditer(i)\n",
    "            for match in matches:\n",
    "                print(match)\n",
    "                print(i)\n",
    "#                     return i\n",
    "                \n",
    "#                 choice1 = input('Enter the path to store the sequences, with a name of the seqres_outfile: ')\n",
    "# #     filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "#                 with open(choice1, 'w') as outfile:\n",
    "#                     for i in seq_list:\n",
    "#                         outfile.write(i)\n",
    "#                         print('Sequences successfully written!')\n",
    "#                         return outfile\n",
    "    \n",
    "    #finds glycosylation sites given G or Y or L then A then P or F or W the TLVGMI as motifs\n",
    "#     filepath = input('Enter the path to the outfile with the seqres sequences: ')\n",
    "#     with open(filepath, 'r') as aa_seq:\n",
    "#         for i in aa_seq:\n",
    "#             pattern = re.compile(r'[GYL]A[PFW][TLVGMI]')\n",
    "#             matches = pattern.finditer(i)\n",
    "#             for match in matches:\n",
    "#                 print(match)     \n",
    "#         return i\n",
    "        \n",
    "    Mainmenu(infile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Please enter the path to where exported pdb files are after being loaded ../Results\n",
      "Please enter the path to where exported pdb files are after being loaded ../Results\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0] myfile3b.pdb\n",
      "\r"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Select pdb file[0-0]:  0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "myfile3b.pdb\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Please enter your choice:  q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program terminated! Goodbye.\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "0",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/njesh/miniconda3/envs/bioinf/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3334: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "Search_file('infile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pdbfile_list():\n",
    "    \"\"\" This function finds all loaded files in the appropriate folder. \"\"\"\n",
    "    \n",
    "    import glob, os\n",
    "    os.chdir(\"../Data\")\n",
    "    file_list = []\n",
    "    for file in glob.glob(\"*.pdb\"):\n",
    "        file_list.append(file)\n",
    "    return file_list\n",
    "\n",
    "def coord_atoms():\n",
    "    \"\"\" This function provides all coordinate atoms in a PDB file, and outputs to a file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    a_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[:4] == 'ATOM':\n",
    "                line_split = line.split()[6:9]\n",
    "                a_list.append(line_split)    \n",
    "    choice1 = input('Enter the name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in a_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!') \n",
    "        \n",
    "def write_backbone_atoms():\n",
    "    \"\"\" This function write out coordinates for all backbone atoms (CA, C, N, O), in PDB file \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    backbone_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[12:16] == \" CA \" and \" O \" and \" N \" and \" C \":\n",
    "                line_split = line.split()[6:9]\n",
    "                backbone_list.append(line_split)\n",
    "    choice1 = input('Enter the name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in backbone_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "def write_CA_atoms():\n",
    "    \"\"\" This function write out coordinates for all alpha-carbon atoms (CA), in a PDB file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    ca_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[:4] == 'ATOM' and line[12:16] == \" CA \":\n",
    "                line_split = line.split()[6:9]\n",
    "                ca_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in ca_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "def write_SEQRES_fasta():\n",
    "    \"\"\" This function extracts seqres sequences in a pdb file and outputs it to a fasta file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Sequences successfully written!')\n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice2 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath2, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "def write_coord_seq():\n",
    "    \"\"\" This function write out the sequence of the PDB file in a fasta format,\n",
    "        based on the coordinates section. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    lis = []\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[:4] == 'ATOM':\n",
    "                line_split = line.split()\n",
    "                lis.append(line_split[3:4])\n",
    "    choice1 = input('Enter name for the output file: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as myfile:\n",
    "        for i in lis:\n",
    "            myfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice2 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath2, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "def write_align_seq():\n",
    "    \"\"\" This function write out a sequence equivalent to the SEQRES section with all the missing\n",
    "        residues converted to ‘X’ characters, in fasta format. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(infile,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Sequences successfully written!')\n",
    "        \n",
    "    choice2 = input('Enter name of the alignment sequence: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    j = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(j, 'r') as fil:\n",
    "        d = {'CYS':'C','ASP':'D','SER':'S','GLN':'Q','LYS':'K','ILE':'I','PRO':'P','THR':'T','PHE':'F','ASN':'N',\n",
    "             'GLY':'G','HIS':'H','LEU':'L','ARG':'R','TRP':'W','TER':'*','ALA':'A','VAL':'V','GLU':'E','TYR':'Y',\n",
    "             'MET':'M','XAA':'X'}\n",
    "        with open(filepath2, 'w') as outf:\n",
    "            for line in fil:\n",
    "                if len(line) %3 == 0:\n",
    "                    upper_seq = line.upper()\n",
    "                    single_seq = ''\n",
    "                    for i in range(int(len(upper_seq)/3)):\n",
    "                        single_seq += d[upper_seq[3*i:3*i+3]]\n",
    "                        outf.write(single_seq)    \n",
    "                    print('Done!')\n",
    "                    return single_seq\n",
    "                else:\n",
    "                    print(\"ERROR: Line was not a factor of 3 in length!\")  \n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice3 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath3 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice3)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath3, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "\n",
    "##sub-options of the main options\n",
    "\n",
    "def submenu1():\n",
    "    \"\"\" This function calls out the sub-menu for writing out coordinate file. \"\"\"\n",
    "    \n",
    "    p = ''\n",
    "    while p == '':\n",
    "        print('\\nS U B  M E N U 1')\n",
    "        print('1. All atom')\n",
    "        print('2. Backbone atoms')\n",
    "        print('3. Alpha carbon atoms only')\n",
    "        print('b. Back')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option.lower() == 'b':\n",
    "            return mainmenu()\n",
    "        elif option == '1':\n",
    "            p = coord_atoms()\n",
    "        elif option == '2':\n",
    "            p = write_backbone_atoms()\n",
    "        elif option == '3':\n",
    "            p = write_CA_atoms()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return p\n",
    "\n",
    "def submenu2():\n",
    "    \"\"\" This function calls out the sub-menu for writing out sequences in a PDB file, in fasta format. \"\"\"\n",
    "    \n",
    "    j = ''\n",
    "    while j == '':\n",
    "        print('\\nS U B  M E N U 2')\n",
    "        print('1. SEQRES sequence')\n",
    "        print('2. Coordinate sequence')\n",
    "        print('3. Alignment sequence')\n",
    "        print('b. Back')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option.lower() == 'b':\n",
    "            return mainmenu()\n",
    "        elif option == '1':\n",
    "            j = write_SEQRES_fasta()\n",
    "        elif option == '2':\n",
    "            j = write_coord_seq()\n",
    "        elif option == '3':\n",
    "            j = write_align_seq()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return j\n",
    "\n",
    "\n",
    "##main-options\n",
    "def mainmenu():\n",
    "    \"\"\" This function calls out the main-menu for writing out PDB files. \"\"\"\n",
    "    \n",
    "    import sys\n",
    "    d = ''\n",
    "    msg = ''    \n",
    "    while d == '':\n",
    "        print('\\nM A I N   M E N U')\n",
    "        print('1. Write out coordinate file')\n",
    "        print('2. Write out sequence(Fasta format)')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option == '1':\n",
    "            msg = 'Option 1'\n",
    "            d = submenu1()\n",
    "        elif option == '2':\n",
    "            msg = 'Option 2'\n",
    "            d = submenu2()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return msg, d    \n",
    "\n",
    "#message, action = mainmenu()\n",
    "\n",
    "#print ('\\nMessage: ', message)\n",
    "#print ('Action: ', action)\n",
    "\n",
    "#the write function\n",
    "\n",
    "def Write_file():\n",
    "    \"\"\" Writes out different portions of PDB files, including atomic coordinates and\n",
    "        protein sequences in fasta format.\n",
    "    \"\"\"\n",
    "    \n",
    "    import os\n",
    "    import sys\n",
    "    pdbfile_list()\n",
    "    \n",
    "    choice = input('Enter name of pdb file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "\n",
    "    file_list = pdbfile_list()\n",
    "    \n",
    "    items = os.listdir('/home/njesh/python-mini-project-JaneNjeri/Data/')\n",
    "    \n",
    "    file_list = [name for name in items if name.endswith('.pdb')]\n",
    "\n",
    "    for count, fileName in enumerate(file_list, 0):\n",
    "        sys.stdout.write(\"[%d] %s\\n\\r\" % (count, fileName))\n",
    "\n",
    "    choice = int(input(\"Select pdb file[0-%s]: \" % count))\n",
    "    print(file_list[choice])\n",
    "            \n",
    "    return choice\n",
    "    choice = infile\n",
    "    mainmenu()\n",
    "    \n",
    "    Mainmenu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pdbfile_list():\n",
    "    \"\"\" This function finds all loaded files in the appropriate folder. \"\"\"\n",
    "    \n",
    "    import glob, os\n",
    "    os.chdir(\"../Data\")\n",
    "    file_list = []\n",
    "    for file in glob.glob(\"*.pdb\"):\n",
    "        file_list.append(file)\n",
    "    return file_list\n",
    "\n",
    "def display_cord_seq():\n",
    "    \"\"\" This function displays the coordinate sequence in a give PDB file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[:4] == 'ATOM':\n",
    "                line_split = line.split()[3:4]\n",
    "                print(line_split)\n",
    "    return file\n",
    "\n",
    "def display_seqres_seq():\n",
    "    \"\"\" This function displays the SEQRES sequences on the console, for a given PDB file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                print(line_split)\n",
    "    return file\n",
    "\n",
    "def display_algn_seq():\n",
    "    \"\"\" This function displays the alignment sequence of a given PDB file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', 'outfile1')\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "    \n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', 'outfile2')\n",
    "    j = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', 'outfile1')\n",
    "    with open(j, 'r') as fil:\n",
    "        d = {'CYS':'C','ASP':'D','SER':'S','GLN':'Q','LYS':'K','ILE':'I','PRO':'P','THR':'T','PHE':'F','ASN':'N',\n",
    "             'GLY':'G','HIS':'H','LEU':'L','ARG':'R','TRP':'W','TER':'*','ALA':'A','VAL':'V','GLU':'E','TYR':'Y',\n",
    "             'MET':'M','XAA':'X'}\n",
    "        with open(filepath2, 'w') as outf:\n",
    "            for line in fil:\n",
    "                if len(line) %3 == 0:\n",
    "                    upper_seq = line.upper()\n",
    "                    single_seq = ''\n",
    "                    for i in range(int(len(upper_seq)/3)):\n",
    "                        single_seq += d[upper_seq[3*i:3*i+3]]\n",
    "                        outf.write(single_seq)    \n",
    "                    return single_seq\n",
    "                else:\n",
    "                    print(\"ERROR: Line was not a factor of 3 in length!\") \n",
    "                    \n",
    "def display_all_nonwater_L():\n",
    "    \"\"\" This function displays all non-water ligands in the protein(if any). \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[:6] == 'HETATM' and line[16:21] != 'HOH':\n",
    "                line_split = line.split()[3:4]\n",
    "                print(line_split)\n",
    "#             else:\n",
    "#                 print('File has no non-water ligands')\n",
    "#                 break\n",
    "\n",
    "def Infor_menu():\n",
    "    \"\"\" This function displays the various options on atomic coordinates and sequences extracted\n",
    "        from a PDB file. \"\"\"\n",
    "    \n",
    "    import sys\n",
    "    d = ''\n",
    "    msg = ''    \n",
    "    while d == '':\n",
    "        print('\\nINFORMATION MENU')\n",
    "        print('1. Display coordinate sequence')\n",
    "        print('2. Display SEQRES sequence')\n",
    "        print('3. Display Alignment sequence')\n",
    "        print('4. Display all non-water ligands in the protein(if any)')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option == '1':\n",
    "            msg = 'Option 1'\n",
    "            d = display_cord_seq()\n",
    "        elif option == '2':\n",
    "            msg = 'Option 2'\n",
    "            d = display_seqres_seq()\n",
    "        elif option == '3':\n",
    "            msg = 'Option 3'\n",
    "            d = display_algn_seq()\n",
    "        elif option == '4':\n",
    "            msg = 'Option 4'\n",
    "            d = display_all_nonwater_L()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return msg, d    \n",
    "\n",
    "# message, action = Infor_menu()\n",
    "\n",
    "# print ('\\nMessage: ', message)\n",
    "# print ('Action: ', action)\n",
    "\n",
    "def Infor_file():\n",
    "    \"\"\" This function Displays information in the PDB files, in terms of the portions defining the file.\n",
    "        Including, coordinate sequence, SEQRES sequence, alignment sequence, all non-water ligands in \n",
    "        the protein (if any).\n",
    "    \"\"\"\n",
    "    \n",
    "    import os\n",
    "    import sys\n",
    "    pdbfile_list()\n",
    "\n",
    "    file_list = pdbfile_list()\n",
    "\n",
    "    items = os.listdir('/home/njesh/python-mini-project-JaneNjeri/PDB_files/')\n",
    "    \n",
    "    file_list = [name for name in items if name.endswith('.pdb')]\n",
    "\n",
    "    for count, fileName in enumerate(file_list, 0):\n",
    "        sys.stdout.write(\"[%d] %s\\n\\r\" % (count, fileName))\n",
    "\n",
    "    choice = int(input(\"Select pdb file[0-%s]: \" % count))\n",
    "    print(file_list[choice])\n",
    "            \n",
    "    return choice\n",
    "    choice = infile\n",
    "    Infor_menu()\n",
    "                \n",
    "    Mainmenu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pdbfile_list():\n",
    "    \"\"\" This function finds all loaded files in the appropriate folder. \"\"\"\n",
    "    import glob, os\n",
    "    os.chdir(\"../Data\")\n",
    "    file_list = []\n",
    "    for file in glob.glob(\"*.pdb\"):\n",
    "        file_list.append(file)\n",
    "    return file_list\n",
    "\n",
    "def extract_SEQRES_seq():\n",
    "    \"\"\" This function extracts seqres sequences in a pdb file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile, to keep sequences for alignment: ')            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Sequences successfully written!')\n",
    "        \n",
    "def Align_2files():\n",
    "    \"\"\" This function aligns two PDB files using local alignment algorithm,\n",
    "        selected by user and outputs the results to a file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    import sys\n",
    "    pdbfile_list()\n",
    "    \n",
    "    file_list = pdbfile_list()\n",
    "    \n",
    "    items = os.listdir('/home/njesh/python-mini-project-JaneNjeri/Data/')\n",
    "    \n",
    "    file_list = [name for name in items if name.endswith('.pdb')]\n",
    "\n",
    "    for count, fileName in enumerate(file_list, 0):\n",
    "        sys.stdout.write('[%d] %s\\n\\r' % (count, fileName))\n",
    "\n",
    "    choice = int(input('Select the first pdb file[0-%s]: ' % count))\n",
    "    print(file_list[choice])\n",
    "    choice2 = int(input('Select the second pdb file[0-%s]: ' % count))\n",
    "    print(file_list[choice2])\n",
    "            \n",
    "    #extracting seqres sequences\n",
    "    print('Please enter the names of the files, starting with the first one.')\n",
    "    choice = extract_SEQRES_seq()\n",
    "\n",
    "    choice2 = extract_SEQRES_seq()\n",
    "            \n",
    "    #building global alignment\n",
    "    from Bio import pairwise2\n",
    "    from Bio.pairwise2 import format_alignment\n",
    "            \n",
    "    choice = input('Enter the name of the first outfile with the extracted sequences: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice)\n",
    "    with open(filepath,'r') as file:\n",
    "        for line in file:\n",
    "            X = line\n",
    "            break\n",
    "                          \n",
    "    choice2 = input('Enter the name of the second outfile with the extracted sequences: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    with open(filepath2,'r') as file:\n",
    "        for line in file:\n",
    "            Y = line\n",
    "            break\n",
    "                    \n",
    "    alignments = pairwise2.align.localxx(X, Y)\n",
    "    for a in alignments:\n",
    "        print(format_alignment(*a))\n",
    "        break\n",
    "#         Z = format_alignment(*a)\n",
    "                \n",
    "    choice3 = input('Enter the name of the outfile with the alignment: ')\n",
    "    filepath3 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice3)\n",
    "    with open(filepath3, 'w') as outfile:\n",
    "        for i in format_alignment(*a):\n",
    "            outfile.writelines(i)\n",
    "            print('All done!')\n",
    "    Mainmenu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Help_file():\n",
    "    \"\"\" This function lists all options and displays a brief description of each. \"\"\"\n",
    "    \n",
    "    print(\"\"\" The available options:\n",
    "\n",
    "        R: Enter name of a PDB file \n",
    "           This option allows the user to read in PDB files for analysis.\n",
    "           \n",
    "        S: Search for PDB files\n",
    "           This option enables the user to search PDB files that are already loaded for analysis.\n",
    "           It also, reports glycosylation sites in the protein with a given pattern for the motif.\n",
    "           \n",
    "        W: Write out PDB files (PDB files in existence)\n",
    "           Writes out different portions of PDB files, including atomic coordinates and\n",
    "           protein sequences in fasta format.\n",
    "           \n",
    "        I: Information on PDB files\n",
    "           Displays information in the PDB files, in terms of the portions defining the file.\n",
    "           Including, coordinate sequence, SEQRES sequence, alignment sequence, all non-water ligands in \n",
    "           the protein (if any).\n",
    "           \n",
    "        A: Alignment\n",
    "           This option extracts SEQRES sequences from two PDB files and aligns the sequences using the local\n",
    "           alignment algorithm.\n",
    "           \n",
    "        H: Help\n",
    "           Lists all options ands a brief description of each option.\n",
    "           \n",
    "        Q: Quit\n",
    "           Terminates the program.\n",
    "        \"\"\")\n",
    "    \n",
    "    Mainmenu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      " ######################################################################\n",
      " # PDB ANALYSIS INTERFACE                                             #\n",
      " ######################################################################\n",
      " # Pick your choice from below                                        #\n",
      " #                                                                    #\n",
      " #  [R] = Enter name of a PDB file                                    #\n",
      " #  [S] = Search for PDB files                                        #\n",
      " #  [W] = Write out PDB files (Files in existence)                    #\n",
      " #  [I] = Information on PDB files                                    #\n",
      " #  [A] = Alignment                                                   #\n",
      " #  [H] = Help                                                        #\n",
      " #  [Q] = Quit                                                        #\n",
      " ######################################################################\n",
      "************MAIN MENU**************\n",
      " Available options \n",
      "                [R]: Enter name of a PDB file \n",
      "                [S]: Search for PDB files\n",
      "                [W]: Write out PDB files (PDB files in existence) \n",
      "                [I]: Information on PDB files\n",
      "                [A]: Alignment\n",
      "                [H]: Help\n",
      "                [Q]: Quit\n",
      "            \n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Please enter your choice:  r\n",
      "Enter name of pdb file:  3b3b.pdb\n",
      "Enter the name of the file:  3b3b.pdb\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AUTHOR': 'A.ADDLAGATTA', 'REMARK': '900'}\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter the name of the file:  3b3b.pdb\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n",
      "File added!\n",
      "File does not exist! Check file again.\n",
      "************MAIN MENU**************\n",
      " Available options \n",
      "                [R]: Enter name of a PDB file \n",
      "                [S]: Search for PDB files\n",
      "                [W]: Write out PDB files (PDB files in existence) \n",
      "                [I]: Information on PDB files\n",
      "                [A]: Alignment\n",
      "                [H]: Help\n",
      "                [Q]: Quit\n",
      "            \n"
     ]
    }
   ],
   "source": [
    "Menu()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
