{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pdbfile_list():\n",
    "    \"\"\" This function finds all loaded files in the appropriate folder. \"\"\"\n",
    "    \n",
    "    import glob, os\n",
    "    os.chdir(\"../Data\")\n",
    "    file_list = []\n",
    "    for file in glob.glob(\"*.pdb\"):\n",
    "        file_list.append(file)\n",
    "    return file_list\n",
    "\n",
    "def coord_atoms():\n",
    "    \"\"\" This function provides all coordinate atoms in a PDB file, and outputs to a file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    a_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[:4] == 'ATOM':\n",
    "                line_split = line.split()[6:9]\n",
    "                a_list.append(line_split)    \n",
    "    choice1 = input('Enter the name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in a_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!') \n",
    "        \n",
    "def write_backbone_atoms():\n",
    "    \"\"\" This function write out coordinates for all backbone atoms (CA, C, N, O), in PDB file \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    backbone_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[12:16] == \" CA \" and \" O \" and \" N \" and \" C \":\n",
    "                line_split = line.split()[6:9]\n",
    "                backbone_list.append(line_split)\n",
    "    choice1 = input('Enter the name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in backbone_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "def write_CA_atoms():\n",
    "    \"\"\" This function write out coordinates for all alpha-carbon atoms (CA), in a PDB file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    ca_list = []\n",
    "    with open(filepath, 'r') as pdb:\n",
    "        for line in pdb:\n",
    "            if line[:4] == 'ATOM' and line[12:16] == \" CA \":\n",
    "                line_split = line.split()[6:9]\n",
    "                ca_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in ca_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "def write_SEQRES_fasta():\n",
    "    \"\"\" This function extracts seqres sequences in a pdb file and outputs it to a fasta file. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(filepath,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Sequences successfully written!')\n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice2 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath2, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "def write_coord_seq():\n",
    "    \"\"\" This function write out the sequence of the PDB file in a fasta format,\n",
    "        based on the coordinates section. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    lis = []\n",
    "    with open(filepath, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[:4] == 'ATOM':\n",
    "                line_split = line.split()\n",
    "                lis.append(line_split[3:4])\n",
    "    choice1 = input('Enter name for the output file: ')\n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as myfile:\n",
    "        for i in lis:\n",
    "            myfile.writelines(i)\n",
    "        print('Done!')\n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice2 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath2, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "def write_align_seq():\n",
    "    \"\"\" This function write out a sequence equivalent to the SEQRES section with all the missing\n",
    "        residues converted to ‘X’ characters, in fasta format. \"\"\"\n",
    "    \n",
    "    import os\n",
    "    choice = input('Enter the name of the file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "    with open(infile,'r') as file:\n",
    "        seq_list = []\n",
    "        for line in file:\n",
    "            if line[:6] == 'SEQRES':\n",
    "                line_split = line.split()[4:]\n",
    "                seq_list.append(line_split)\n",
    "    choice1 = input('Enter name of the outfile: ')            \n",
    "    filepath1 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(filepath1, 'w') as outfile:\n",
    "        for i in seq_list:\n",
    "            outfile.writelines(i)\n",
    "        print('Sequences successfully written!')\n",
    "        \n",
    "    choice2 = input('Enter name of the alignment sequence: ')\n",
    "    filepath2 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    j = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice1)\n",
    "    with open(j, 'r') as fil:\n",
    "        d = {'CYS':'C','ASP':'D','SER':'S','GLN':'Q','LYS':'K','ILE':'I','PRO':'P','THR':'T','PHE':'F','ASN':'N',\n",
    "             'GLY':'G','HIS':'H','LEU':'L','ARG':'R','TRP':'W','TER':'*','ALA':'A','VAL':'V','GLU':'E','TYR':'Y',\n",
    "             'MET':'M','XAA':'X'}\n",
    "        with open(filepath2, 'w') as outf:\n",
    "            for line in fil:\n",
    "                if len(line) %3 == 0:\n",
    "                    upper_seq = line.upper()\n",
    "                    single_seq = ''\n",
    "                    for i in range(int(len(upper_seq)/3)):\n",
    "                        single_seq += d[upper_seq[3*i:3*i+3]]\n",
    "                        outf.write(single_seq)    \n",
    "                    print('Done!')\n",
    "                    return single_seq\n",
    "                else:\n",
    "                    print(\"ERROR: Line was not a factor of 3 in length!\")  \n",
    "        \n",
    "    with open(choice, 'r') as myfile:\n",
    "        header = ''\n",
    "        for line in myfile:\n",
    "            if line.startswith(\"TITLE\"): \n",
    "                head_split = line.split()\n",
    "                header = header + ' '.join(head_split[1:])\n",
    "            \n",
    "    choice3 = input('Enter output file name with a .fasta extension: ')\n",
    "    filepath3 = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice3)\n",
    "    z = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Results', choice2)\n",
    "    with open(z, 'r') as file:\n",
    "        with open(filepath3, 'w') as output:\n",
    "            for i in file:\n",
    "                output.writelines('>' + header + '\\n' + i)\n",
    "        print('Fasta file generated!')\n",
    "        \n",
    "\n",
    "##sub-options of the main options\n",
    "\n",
    "def submenu1():\n",
    "    \"\"\" This function calls out the sub-menu for writing out coordinate file. \"\"\"\n",
    "    \n",
    "    p = ''\n",
    "    while p == '':\n",
    "        print('\\nS U B  M E N U 1')\n",
    "        print('1. All atom')\n",
    "        print('2. Backbone atoms')\n",
    "        print('3. Alpha carbon atoms only')\n",
    "        print('b. Back')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option.lower() == 'b':\n",
    "            return mainmenu()\n",
    "        elif option == '1':\n",
    "            p = coord_atoms()\n",
    "        elif option == '2':\n",
    "            p = write_backbone_atoms()\n",
    "        elif option == '3':\n",
    "            p = write_CA_atoms()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return p\n",
    "\n",
    "def submenu2():\n",
    "    \"\"\" This function calls out the sub-menu for writing out sequences in a PDB file, in fasta format. \"\"\"\n",
    "    \n",
    "    j = ''\n",
    "    while j == '':\n",
    "        print('\\nS U B  M E N U 2')\n",
    "        print('1. SEQRES sequence')\n",
    "        print('2. Coordinate sequence')\n",
    "        print('3. Alignment sequence')\n",
    "        print('b. Back')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option.lower() == 'b':\n",
    "            return mainmenu()\n",
    "        elif option == '1':\n",
    "            j = write_SEQRES_fasta()\n",
    "        elif option == '2':\n",
    "            j = write_coord_seq()\n",
    "        elif option == '3':\n",
    "            j = write_align_seq()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return j\n",
    "\n",
    "\n",
    "##main-options\n",
    "def mainmenu():\n",
    "    \"\"\" This function calls out the main-menu for writing out PDB files. \"\"\"\n",
    "    \n",
    "    import sys\n",
    "    d = ''\n",
    "    msg = ''    \n",
    "    while d == '':\n",
    "        print('\\nM A I N   M E N U')\n",
    "        print('1. Write out coordinate file')\n",
    "        print('2. Write out sequence(Fasta format)')\n",
    "        print('q. Quit')\n",
    "        option = input('Select an option: ')\n",
    "        if option.lower() == 'q':\n",
    "            sys.exit()\n",
    "        elif option == '1':\n",
    "            msg = 'Option 1'\n",
    "            d = submenu1()\n",
    "        elif option == '2':\n",
    "            msg = 'Option 2'\n",
    "            d = submenu2()\n",
    "        else:\n",
    "            print ('Invalid selection!')\n",
    "    return msg, d    \n",
    "\n",
    "message, action = mainmenu()\n",
    "\n",
    "print ('\\nMessage: ', message)\n",
    "print ('Action: ', action)\n",
    "\n",
    "#the write function\n",
    "\n",
    "def Write_file():\n",
    "    \"\"\" Writes out different portions of PDB files, including atomic coordinates and\n",
    "        protein sequences in fasta format.\n",
    "    \"\"\"\n",
    "    \n",
    "    import os\n",
    "    import sys\n",
    "    pdbfile_list()\n",
    "    \n",
    "    choice = input('Enter name of pdb file: ')\n",
    "    filepath = os.path.join('/home/njesh/python-mini-project-JaneNjeri/Data', choice)\n",
    "\n",
    "    file_list = pdbfile_list()\n",
    "    \n",
    "    items = os.listdir('/home/njesh/python-mini-project-JaneNjeri/Data/')\n",
    "    \n",
    "    file_list = [name for name in items if name.endswith('.pdb')]\n",
    "\n",
    "    for count, fileName in enumerate(file_list, 0):\n",
    "        sys.stdout.write(\"[%d] %s\\n\\r\" % (count, fileName))\n",
    "\n",
    "    choice = int(input(\"Select pdb file[0-%s]: \" % count))\n",
    "    print(file_list[choice])\n",
    "            \n",
    "    return choice\n",
    "    choice = infile\n",
    "    mainmenu()\n",
    "    \n",
    "    Mainmenu()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
